pydrive2/fs.py

```diff
@@ -0,0 +1,476 @@
+                if item["mimeType"] == FOLDER_MIME_TYPE:
+                    new_query_ids[item["id"]] = item_path
+                    self._cache_path_id(item_path, item["id"])
+                    continue
+
+                contents.append(
+                    {
+                        "name": posixpath.join(bucket, item_path),
+                        "type": "file",
+                        "size": int(item["fileSize"]),
+                        "checksum": item["md5Checksum"],
+                    }
+                )
+
+        if detail:
+            return contents
+        else:
+            return [content["name"] for content in contents]
+
+    def upload_fobj(self, stream, rpath, callback=None, **kwargs):
+        parent_id = self._get_item_id(self._parent(rpath), create=True)
+        if callback:
+            stream = CallbackIOWrapper(
+                callback.relative_update, stream, "read"
+            )
+        return self.gdrive_upload_fobj(
+            posixpath.basename(rpath.rstrip("/")), parent_id, stream
+        )
+
+    def put_file(self, lpath, rpath, callback=None, **kwargs):
+        if callback:
+            callback.set_size(os.path.getsize(lpath))
+        with open(lpath, "rb") as stream:
+            self.upload_fobj(stream, rpath, callback=callback)
+
+    @_gdrive_retry
+    def gdrive_upload_fobj(self, title, parent_id, stream, callback=None):
+        item = self.client.CreateFile(
+            {"title": title, "parents": [{"id": parent_id}]}
+        )
+        item.content = stream
+        item.Upload()
+        return item
+
+    def cp_file(self, lpath, rpath, **kwargs):
+        """In-memory streamed copy"""
+        with self.open(lpath) as stream:
+            # IterStream objects doesn't support full-length
+            # seek() calls, so we have to wrap the data with
+            # an external buffer.
+            buffer = io.BytesIO(stream.read())
+            self.upload_fobj(buffer, rpath)
+
+    def get_file(self, lpath, rpath, callback=None, **kwargs):
+        item_id = self._get_item_id(lpath)
+        return self.gdrive_get_file(item_id, rpath, callback)
+
+    @_gdrive_retry
+    def gdrive_get_file(self, item_id, rpath, callback):
+        param = {"id": item_id}
+        # it does not create a file on the remote
+        gdrive_file = self.client.CreateFile(param)
+
+        extra_args = {}
+        if callback:
+            gdrive_file.FetchMetadata(fields="fileSize")
+            callback.set_size(int(gdrive_file.get("fileSize")))
+            extra_args["callback"] = callback.update
+
+        gdrive_file.GetContentFile(rpath, **extra_args)
+
+    def _open(self, path, mode, **kwargs):
+        assert mode in {"rb", "wb"}
+        if mode == "wb":
+            return GDriveBufferedWriter(self, path)
+        else:
+            item_id = self._get_item_id(path)
+            return self.gdrive_open_file(item_id)
+
+    @_gdrive_retry
+    def gdrive_open_file(self, item_id):
```

Comment:
question - what happened to the `contextmanager` open? will we keep it in DVC?

Reply:
> question - what happened to the contextmanager open? will we keep it in DVC?

`_open()` itself will be wrapped by `AbstractFileSystem.open` which is a context manager.

Reply:
👍